packages/compiler/core/helpers/type-name-utils.ts

```diff
@@ -53,6 +53,10 @@ export function getTypeName(type: Type, options?: TypeNameOptions): string {
       return type.name;
   }
 
+  if (typeof type !== "object") {
```

Comment:
when is this getting hit?

Reply:
In the scenario Laurent posted.

Reply:
then there is a bug somewhere else, something is cheating it way into calling this function.

Reply:
if you call `getTypeName("float32")` typescript will complain

Reply:
Yes, originally I tried calling `getTypeName(type)` and it exploded.

Reply:
my point is that it is the wrong place to fix it. 
![image](https://github.com/microsoft/typespec/assets/1031227/e1eb8652-76ed-47c3-a20e-19d27652c13e)
By this logic we should be checking every argument from every function again in case someone cheat the value in.

Reply:
So, here's the signature for `$encode`
```
// signature
export function $encode(
  context: DecoratorContext,
  target: Scalar | ModelProperty,
  encoding: string | EnumMember,
  encodeAs?: Scalar
)

// extern dec 
extern dec encode(
  target: Scalar | ModelProperty,
  encoding: string | EnumMember,
  encodedAs?: string | numeric
)
```

And so the issue is that inside of `$encode`, the `encodeAs` property is `int32`, not a  `Type` object.

Reply:
Without the typeof check, it will return this as type `(unnamed type)`

Reply:
yeah so here to rpoblem is that `encodeAs` says it is a scalar but it isn't necessarly. I see 2 options:
1. we change the extrern dec to be `encodeAs?: Reflection.Scalar`
2. first thing in $encode is to verify it is a scalar again.

This is why it would be awesome to have the TS signature generated from the TSP declaration that way it would force the TYpescript to be accurate and deal with all the things it could receive.

Reply:
you mean if you do `opt 1`? this is a pattern we have everywhere we we expect an exact  TypeSpec type, maybe that message could be updated too to be clearer that we expecting exactly a Scalar and not a subtype of it.

Reply:
one downside of opt1 is you can now pass other scalar that are not numeric or string like duration or utcDateTime which don't really make much sense as an encode target.

Reply:
Yeah, that's opt 1.

Reply:
But even with opt2... the string "int32" is a scalar... it's just not the scalar type? I feel like the original error message "Encoding 'seconds' cannot be used on type 's'. Expected 'numeric', but got 'string'" is correct, but without the object check, I don't see any way to get that outcome unless I somehow convert `"int32"` into a Scalar object.